google/cloud/spanner/value.cc

```diff
@@ -0,0 +1,173 @@
+}
+
+Value::Value(double v) {
+  type_.set_code(google::spanner::v1::TypeCode::FLOAT64);
+  if (std::isnan(v)) {
+    value_.set_string_value("NaN");
+  } else if (std::isinf(v)) {
+    value_.set_string_value(v < 0 ? "-Infinity" : "Infinity");
+  } else {
+    value_.set_number_value(v);
+  }
+}
+
+Value::Value(std::string v) {
+  type_.set_code(google::spanner::v1::TypeCode::STRING);
+  value_.set_string_value(std::move(v));
+}
+
+bool operator==(Value a, Value b) {
+  if (a.type_.code() != b.type_.code()) return false;
+  if (a.is_null() != b.is_null()) return false;
+  if (a.is_null()) return true;  // They are both null
+  switch (a.type_.code()) {
+    case google::spanner::v1::TypeCode::BOOL:
+      return *a.get<bool>() == *b.get<bool>();
+    case google::spanner::v1::TypeCode::INT64:
+      return *a.get<std::int64_t>() == *b.get<std::int64_t>();
+    case google::spanner::v1::TypeCode::FLOAT64:
+      return *a.get<double>() == *b.get<double>();
+    case google::spanner::v1::TypeCode::STRING:
+      return *a.get<std::string>() == *b.get<std::string>();
+    default:
+      return false;
+  }
+}
+
+std::ostream& operator<<(std::ostream& os, Value v) {
+  auto const& code = v.type_.code();
+  auto const* const descriptor = google::spanner::v1::TypeCode_descriptor();
+  os << "Value{" << descriptor->FindValueByNumber(code)->name() << ":";
+  if (v.is_null()) return os << "(null)}";
+  switch (code) {
+    case google::spanner::v1::TypeCode::BOOL:
+      os << *v.get<bool>();
+      break;
+    case google::spanner::v1::TypeCode::INT64:
+      os << *v.get<std::int64_t>();
+      break;
+    case google::spanner::v1::TypeCode::FLOAT64:
+      os << *v.get<double>();
+      break;
+    case google::spanner::v1::TypeCode::STRING:
+      os << *v.get<std::string>();
+      break;
+  }
+  return os << "}";
+}
+
+bool Value::is_null() const {
+  return value_.kind_case() == google::protobuf::Value::kNullValue;
+}
+
+//
+// Value::IsType
+//
+
+bool Value::IsType(bool) const {
+  return type_.code() == google::spanner::v1::TypeCode::BOOL;
+}
+
+bool Value::IsType(std::int64_t) const {
+  return type_.code() == google::spanner::v1::TypeCode::INT64;
+}
+
+bool Value::IsType(double) const {
+  return type_.code() == google::spanner::v1::TypeCode::FLOAT64;
+}
+
+bool Value::IsType(std::string) const {
+  return type_.code() == google::spanner::v1::TypeCode::STRING;
+}
+
+//
+// Value::GetValue
+//
+
+StatusOr<bool> Value::GetValue(bool) const {
+  auto const status = CheckValidity<bool>(*this);
+  if (!status.ok()) return status;
+  return value_.bool_value();
+}
+
+StatusOr<std::int64_t> Value::GetValue(std::int64_t) const {
+  auto const status = CheckValidity<std::int64_t>(*this);
+  if (!status.ok()) return status;
+  auto const& s = value_.string_value();
+  std::size_t processed = 0;
+  long long x = std::stoll(s, &processed, 10);
+  if (processed != s.size()) {
+    GCP_LOG(FATAL) << "Failed to parse number from string: \"" << s << "\"";
```

Comment:
FYI: Right now `GCP_LOG(FATAL)` does not really terminate, we should probably fix that.

Reply:
Ahh, ok. well, that's probably fine for now.